docs/src/utility_calculators.md

```diff
@@ -0,0 +1,97 @@
+
+[CombinationCalculator](@ref) has an additional feature that allows it to edit what keywords are passed down on the calculators. This allows e.g. generating data that can be used by all the calculators.
+
+The default is passing keywords as they are. There are two ways to change this, fist is global setting with [generate_keywords](@ref) function or by givin keyword `keywordgenerator`. Global option is intended e.g. for package authors to make global settings. You could e.g. use it to calculate pair list for pair potentials:
+
+```julia
+function AtomsCalculators.UntilityCalculators.generate_keywords(
+    sys, 
+    pp::PairPotential...; 
+    kwargs...
+)
+    cutoff = cutoff_radius(pp[begin])
+    if all( c ->  cutoff_radius(c) ≈ cutoff, pp )
+        plist = PairList(sys, cutoff)
+        return (; :plist => plist, kwargs...)
+    else
+        return kwargs
+    end
+end
+```
+
+PairPotential author could have this in his package and it would set it globally.
+
+`keyword_generator` keyword allows you to set keywords for individual `CombinationCalculator`. 
+
+```julia
+function remove_keywords(sys, pp...; kwargs...)
+    return (; )
+end
+
+CombinationCalculator(calc1, calc2; keyword_generator=remove_keywords)
+```
+
+
+## Reporting Calculator
+
+[ReportingCalculator](@ref) is used to collect information during calculation. To do this ReportingCalculator call a function to generate a message and then passes that message to a `Channel`.
+
+By default message is generated by a function [generate_message](@ref), but this can be changed by giving a keyword `message_function`. The default message is the value of latest calculation
+
+```julia
+c = Channel(32)
+rep_calc = ReportingCalculator(calc, c)
+
+e = AtomsCalculators.potential_energy(system, rep_calc)
+e == fetch(c)         # true
+e == take!(rep_calc)  # true
+```
+
+You can call `fetch` and `take!` directly to [ReportingCalculator](@ref) to access the `Channel`. 
+
+```julia
+function my_message(sys, calc, e::Number; kwords...)
+    return FlexibleSystem(sys; energy=e)
+end
+
+rep_calc = ReportingCalculator(calc, Channel(32); message_function=my_message)
+
+f = AtomsCalculators.forces(system, rep_calc)  # no message generated
+e = AtomsCalculators.potential_energy(system, rep_calc) # message generated
+```
+
+
+## Subsystem Calculator
+
+[SubSystemCalculator](@ref) is used to split input system into smaller parts. This is useful e.g. for QM/MM calculations.
+
+To use [SubSystemCalculator](@ref) you need to give an other calculator and indexes that define subsystem. E.g., if your subsystem is defined by atoms 1 to 10 and 15 you can create subsystem calculator by
+
+```julia
+sub_calc = SubSystemCalculator( mycalc, [1:10..., 15])
+```
+
+This calculator will then create a system with atoms 1 to 10 and 15 and pass that one into the calculator. Zero forces are given to all other atoms.
```

Comment:
As I already expressed in the zoom call, I think we should be very careful with this pattern of wrapping other calculators to obtain functionality. In my opinion this will make it harder down the road to define specialised methods for specific calculators. My main gripes is for example default parameters or choices in the algorithm for "cheap" calculators like empirical potentials versus "expensive" calculators like DFT computations. If you keep wrapping you loose the elegance of Julia to specialise behaviour by just implementing a new method for a specific calculator type.

Since in the discussion it became clear, that both the reporting as well as the subsystem calculator are needed for QMMM, I'd suggest to only implement a QMMM calculator and worry about coming up with more general primitives like how to extract parts of the data (reporting) or how to split into subsystems later as we have more understanding of what our ecosystem looks like.

Reply:
This goes on what do we mean by QM/MM calculation. If the definition is one QM and one MM system. Then making separate QM/MM calculator is the way to go.

I am however thinking of much more complicated calculators, like one QM, one ML and one MM, or several ML and QM systems. In these cases the simple QM/MM calculator wont work.

Basicly these 3 calculators (Combination, Subsystem, and Reporting) are designed to make this kind of complicated calculations easily doable.

There is of course the issue you are pointing. But, I am not sure is using dispatch a correct way to handle it. We could e.g. implement a "calculation_hit" function that would allow upstream get info on, what to expect from the calculator.